esp32/machine_hw_spi: Use a 2 item SPI queue for long transfers.

jonathanhogg opened the pull request:
Using a 2-item transaction queue instead of 1 allows long transfers to be executed with the minimum inter-transaction delay. Limit maximum transaction length to ensure an integer multiple of the SPI `bits` setting are transferred. Fixes #7511.

dpgeorge commented:
This change looks pretty good to me.  If I understand it correctly then anyone using SPI with `bits=` 8, 16, 24, 32 will see no change if byte length <= 4092, and will see reduced gaps between chunks if byte length > 4092.

For other multiples of `bits`, the gap between chunks will now occur at a slightly different location (a `bits` boundary), and the gap will be smaller.

So +1

dpgeorge commented on ports/esp32/machine_hw_spi.c:
```diff
@@ -269,6 +269,16 @@ STATIC void machine_hw_spi_deinit(mp_obj_base_t *self_in) {
     }
 }
 
+STATIC mp_uint_t gcd(mp_uint_t x, mp_uint_t y) {
```
can you please make this a loop rather than recursive, just in case the compiler cannot eliminate the tail recursion

jonathanhogg replied:
Sure. I was being lazy here and just grabbing the first/simplest GCD implementation I could find. I wasn't sure if I'd be instructed to actually put it somewhere else as a utility function.

dpgeorge commented on ports/esp32/machine_hw_spi.c:
```diff
@@ -213,7 +213,7 @@ STATIC void machine_hw_spi_init_internal(
         .mode = self->phase | (self->polarity << 1),
         .spics_io_num = -1, // No CS pin
-        .queue_size = 1,
+        .queue_size = 2,
```
what does this actually do, does it allocate internal memory for both queues?  what about the `len <= 4` case in `machine_hw_spi_transfer` below, is that ok that it still uses 1 queue?

jonathanhogg replied:
This is a good question. The IDF documentation just says:

> Transaction queue size. This sets how many transactions can be ‘in the air’ (queued using `spi_device_queue_trans` but not yet finished using `spi_device_get_trans_result`) at the same time.

I am assuming that it something to do with memory allocation, but I assume the queue is just pointers to the transaction structures, so I can't imagine it being that big a deal.

For small transfers, these still count as a single transaction item in the queue, it's just that the bits are stored straight in the transaction instead of being a pointer. So, yes, small transfers will still use 1 queue item.

dpgeorge replied:
Ok!

jonathanhogg commented:
> This change looks pretty good to me. If I understand it correctly then anyone using SPI with `bits=` 8, 16, 24, 32 will see no change if byte length <= 4092, and will see reduced gaps between chunks if byte length > 4092.

Yes. In fact it's kind of worse with the current implementation for long transfers as the inter-transaction delays can cross a word boundary so you might have been half way through one of those 16/24/32 bits and suddenly been hit with a huge gap.

> For other multiples of `bits`, the gap between chunks will now occur at a slightly different location (a `bits` boundary), and the gap will be smaller.
> 
> So +1

Yes, odd word sizes will have a slightly reduced maximum single transaction size, but shorter gaps for long transfers and always between words.

dpgeorge commented:
>  In fact it's kind of worse with the current implementation for long transfers as the inter-transaction delays can cross a word boundary so you might have been half way through one of those 16/24/32 bits and suddenly been hit with a huge gap.

Is that true? 8/16/24/32 all divide evenly into 4092*8 bits.

jonathanhogg commented:
> Is that true? 8/16/24/32 all divide evenly into 4092*8 bits.

Oh yes, sorry, that is of course true! It's only odd word sizes that would have caused problems.

jonathanhogg commented:
Actually, hold off on this for a moment: I was just testing with the len <= 4 case and I'm getting IDF errors – even though I've not touched that code.

jonathanhogg committed:
esp32/machine_hw_spi: Use a 2 item SPI queue for long transfers.

Using a 2-item transaction queue instead of 1 allows long transfers to 
be executed with the minimum inter-transaction delay. Limit maximum 
transaction length to ensure an integer multiple of the SPI `bits` 
setting are transferred. Fixes #7511.

jonathanhogg commented:
OK. False alarm. I keep forgetting that I'm using `bits=72`. I tried to send a 3 byte buffer and it threw errors. Digging into this it's – of course – because the length is rounded down to a multiple of `bits`, which turned out to be `0`. Trying to send 0 bits causes Bad Things to happen.

I've added a check for this and am throwing a `ValueError`. Just returning without executing a transfer would be semantically correct, but I felt that this is almost certain to be a mistake and something that the user would want to know about. This is a code path that most people wouldn't get near I don't think. I'm happy to change it to a return if you think that's more in-keeping with other functionality.

dpgeorge commented:
> I've added a check for this and am throwing a `ValueError`.

That looks OK.  I don't think any other port has a specific check for zero length and it may be undefined behaviour on them.  Can revisit this if needed in the future.
